vllm/core/block_manager.py

```diff
@@ -15,40 +64,86 @@ class BlockAllocator:
     the reference count becomes zero, the block is added back to the free list.
     """
 
-    def __init__(
-        self,
-        device: Device,
-        block_size: int,
-        num_blocks: int,
-    ) -> None:
+    def __init__(self,
+                 device: Device,
+                 block_size: int,
+                 num_blocks: int,
+                 eviction_policy: EvictionPolicy = EvictionPolicy.LRU) -> None:
         self.device = device
         self.block_size = block_size
         self.num_blocks = num_blocks
 
-        # Initialize the free blocks.
-        self.free_blocks: BlockTable = []
-        for i in range(num_blocks):
-            block = PhysicalTokenBlock(device=device,
-                                       block_number=i,
-                                       block_size=block_size)
-            self.free_blocks.append(block)
-
-    def allocate(self) -> PhysicalTokenBlock:
-        if not self.free_blocks:
-            raise ValueError("Out of memory! No free blocks are available.")
-        block = self.free_blocks.pop()
-        block.ref_count = 1
+        self.eviction_policy = eviction_policy
+
+        self.current_num_blocks = 0
+        self.table: Dict[int, PhysicalTokenBlock] = {}
+        self.free_table: Dict[int, PhysicalTokenBlock] = {}
+
+    def evict(self) -> PhysicalTokenBlock:
+        if self.eviction_policy == EvictionPolicy.LRU:
+            return lru_eviction(self.free_table)
+        else:
+            raise ValueError(
+                f"Unknown cache eviction policy: {self.eviction_policy}")
+
+    def allocate_block(self, block_hash: int,
+                       prefix_len: int) -> PhysicalTokenBlock:
+        if self.current_num_blocks == self.num_blocks:
+            block = self.evict()
+            block.block_hash = block_hash
+            block.prefix_len = prefix_len
+            return block
+        block = PhysicalTokenBlock(device=self.device,
+                                   block_number=self.current_num_blocks,
+                                   block_size=self.block_size,
+                                   block_hash=block_hash,
+                                   prefix_len=prefix_len)
+        self.current_num_blocks += 1
+        return block
+
+    def allocate(self,
+                 block_hash: Optional[int] = None,
+                 prefix_len: int = 0) -> PhysicalTokenBlock:
+        if block_hash is None:
+            block_hash = monotonic()
+        if block_hash in self.free_table:
+            assert block_hash not in self.table
+            block = self.free_table[block_hash]
+            assert block.ref_count == 0
+            self.table[block_hash] = block
+            block.ref_count += 1
+            del self.free_table[block_hash]
+            assert block.block_hash == block_hash
+            return block
```

Comment:
Correct. This way, we can obtain the eviction candidate faster. Instead of iterating over all allocated blocks, we only iterate over blocks that have a ref count of 0.